src/encryption.rs

```diff
@@ -0,0 +1,798 @@
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    match key_manager.delete_file(fname) {
+        Ok(()) => ptr::null(),
+        Err(err) => unsafe {
+            libc::strdup(
+                CString::new(format!(
+                    "Encryption key manager delete file failure: {}",
+                    err
+                ))
+                .unwrap()
+                .into_raw(),
+            )
+        },
+    }
+}
+
+extern "C" fn encryption_key_manager_link_file(
+    ctx: *mut c_void,
+    src_fname: *const c_char,
+    dst_fname: *const c_char,
+) -> *const c_char {
+    let key_manager = unsafe { &**(ctx as *mut *const dyn EncryptionKeyManager) };
+    let src_fname = match unsafe { CStr::from_ptr(src_fname).to_str() } {
+        Ok(ret) => ret,
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    let dst_fname = match unsafe { CStr::from_ptr(dst_fname).to_str() } {
+        Ok(ret) => ret,
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
+            };
+        }
+    };
+    match key_manager.link_file(src_fname, dst_fname) {
+        Ok(()) => ptr::null(),
+        Err(err) => unsafe {
+            libc::strdup(
+                CString::new(format!(
+                    "Encryption key manager delete file failure: {}",
+                    err
+                ))
+                .unwrap()
+                .into_raw(),
+            )
+        },
+    }
+}
+
+extern "C" fn encryption_key_manager_rename_file(
+    ctx: *mut c_void,
+    src_fname: *const c_char,
+    dst_fname: *const c_char,
+) -> *const c_char {
+    let key_manager = unsafe { &**(ctx as *mut *const dyn EncryptionKeyManager) };
+    let src_fname = match unsafe { CStr::from_ptr(src_fname).to_str() } {
+        Ok(ret) => ret,
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
```

Comment:
It looks like leaking. 
1. CString::into_raw() consumes the CString and transfers ownership of the string to a C caller.
2. strdup allocates and duplicates the string of into_raw().

So 1 is leaked.

Reply:
The pointer is received by C caller and deleted. See, for example, crocksdb_encryption_key_manager_impl_t::RenameFile in c.cc.

Reply:
good catch! fixed.